<a href="https://colab.research.google.com/github/Tirta2241720045/Machine_Learning/blob/main/Minggu6/Week_6_Praktikum.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Job Sheet 6: Ensemble Learning
# Nama: Tirta Nurrochman Bintang Prawira
# NIM: 2241720045
# Kelas: TI-3A
# No: 27

# Praktikum 1 Bagging dengan RandomForest

Import Library

In [1]:
import numpy as np
import pandas as pd
from sklearn.tree import DecisionTreeClassifier # import DT
from sklearn.ensemble import RandomForestClassifier # import RandomForest
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report

1. **Mengimpor Library**:
   - `numpy` dan `pandas` digunakan untuk manipulasi data.
   - `DecisionTreeClassifier` dan `RandomForestClassifier` untuk model klasifikasi.

2. **Pembagian Dataset**:
   - `train_test_split` membagi data menjadi *training* dan *testing* set.

3. **Evaluasi Model**:
   - `accuracy_score` dan `classification_report` untuk mengukur performa model.


Persiapan Data

In [2]:
# Load data
df = pd.read_csv('/content/drive/MyDrive/dataset/wbc (1).csv')

df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

1. **`df = pd.read_csv('/content/drive/MyDrive/dataset/wbc (1).csv')`**:  
   Membaca dataset dari file CSV (`wbc (1).csv`) dan menyimpannya ke dalam DataFrame `df` untuk analisis lebih lanjut.

2. **`df.head()`**:  
   Menampilkan 5 baris pertama dari DataFrame `df` untuk memberikan gambaran awal tentang struktur data (kolom dan beberapa contoh baris).


In [3]:
# Cek kolom null
df.isnull().sum()

id                           0
diagnosis                    0
radius_mean                  0
texture_mean                 0
perimeter_mean               0
area_mean                    0
smoothness_mean              0
compactness_mean             0
concavity_mean               0
concave points_mean          0
symmetry_mean                0
fractal_dimension_mean       0
radius_se                    0
texture_se                   0
perimeter_se                 0
area_se                      0
smoothness_se                0
compactness_se               0
concavity_se                 0
concave points_se            0
symmetry_se                  0
fractal_dimension_se         0
radius_worst                 0
texture_worst                0
perimeter_worst              0
area_worst                   0
smoothness_worst             0
compactness_worst            0
concavity_worst              0
concave points_worst         0
symmetry_worst               0
fractal_dimension_worst      0
Unnamed: 32                569
dtype: int64

1. **`df.isnull().sum()`**:
   - Mengecek setiap kolom apakah terdapat nilai `NaN` (null/missing values).
   - Menghasilkan jumlah nilai kosong per kolom.


In [4]:
# Seleksi fitur

# Slice dataframe mulai dari kolom 'radius_mean' sampai 'fractal_dimension_worst'
X = df.iloc[:,3:-1]
y = df['diagnosis']
y = y.map({'M':1, 'B':0}) # Encode label

# Cek jumlah fitur dan instance
X.shape

(569, 29)

1. **`X = df.iloc[:, 3:-1]`**:  
   Mengambil subset data `df` dari kolom ke-4 (`radius_mean`) hingga kolom sebelum yang terakhir, untuk dijadikan fitur (`X`).

2. **`y = df['diagnosis']`**:  
   Mengambil kolom `diagnosis` sebagai label (`y`).

3. **`y = y.map({'M':1, 'B':0})`**:  
   Mengubah label `M` (Malignant) menjadi 1 dan `B` (Benign) menjadi 0 agar dapat digunakan dalam model klasifikasi.

4. **`X.shape`**:  
   Menampilkan jumlah fitur (kolom) dan instance (baris) dalam `X`.


Split data training dan testing

In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

1. **`train_test_split(X, y, test_size=0.2, random_state=1)`**:  
   Membagi data fitur (`X`) dan label (`y`) menjadi *training* dan *testing* set:
   - `test_size=0.2`: 20% data digunakan sebagai *testing set*, sisanya 80% untuk *training set*.
   - `random_state=1`: Mengatur *random seed* agar pembagian data konsisten setiap kali dijalankan.

2. **`X_train, X_test, y_train, y_test`**:  
   - `X_train` dan `y_train`: Data untuk melatih model.
   - `X_test` dan `y_test`: Data untuk menguji performa model.



Traning Decision Tree

In [6]:
# Secara default, DecisionTreeClassifier dari scikit-learn akan menggunakan nilai "Gini" untuk kriteria
# Terdapat beberapa "hyperparamater" yang dapat digunakan. Silahka baca dokumentasi
# Pada kasus ini kita akan menggunakan parameter default
dt = DecisionTreeClassifier()

# Sesuaikan dt ke set training
dt.fit(X_train, y_train)

# Memprediksi label set test
y_pred_dt = dt.predict(X_test)

#  menghitung set accuracy
acc_dt = accuracy_score(y_test, y_pred_dt)
print("Test set accuracy: {:.2f}".format(acc_dt))
print(f"Test set accuracy: {acc_dt}")

Test set accuracy: 0.94
Test set accuracy: 0.9385964912280702


1. **`dt = DecisionTreeClassifier()`**:  
   Inisialisasi model *Decision Tree* dengan parameter default (menggunakan kriteria Gini).

2. **`dt.fit(X_train, y_train)`**:  
   Melatih model *Decision Tree* menggunakan *training set* (`X_train`, `y_train`).

3. **`y_pred_dt = dt.predict(X_test)`**:  
   Memprediksi label untuk *testing set* (`X_test`) berdasarkan model yang telah dilatih.

4. **`acc_dt = accuracy_score(y_test, y_pred_dt)`**:  
   Menghitung akurasi prediksi model pada *testing set* dengan membandingkan `y_pred_dt` (prediksi) dan `y_test` (label asli).

5. **`print(...)`**:  
   Menampilkan akurasi model dalam format desimal (2 tempat) dan nilai akurasinya.


Training RandomForest

In [7]:
# Pada kasus kali ini kita akan menggunakan estimator pada RandomForest
# Untuk detail parameter (hyperparameter) silahkan cek dokumentasi

rf = RandomForestClassifier(n_estimators=10, random_state=1)

# Sesuaikan dt ke set training
rf.fit(X_train, y_train)

# Memprediksi label set test
y_pred_rf = rf.predict(X_test)

#  menghitung set accuracy
acc_rf = accuracy_score(y_test, y_pred_rf)
print("Test set accuracy: {:.2f}".format(acc_rf))
print(f"Test set accuracy: {acc_rf}")

Test set accuracy: 0.96
Test set accuracy: 0.956140350877193


1. **`rf = RandomForestClassifier(n_estimators=10, random_state=1)`**:  
   Inisialisasi model *Random Forest*:
   - `n_estimators=10`: Menggunakan 10 *decision tree* dalam *Random Forest*.
   - `random_state=1`: Mengatur *random seed* agar hasilnya konsisten setiap kali dijalankan.

2. **`rf.fit(X_train, y_train)`**:  
   Melatih model *Random Forest* menggunakan *training set* (`X_train`, `y_train`).

3. **`y_pred_rf = rf.predict(X_test)`**:  
   Memprediksi label untuk *testing set* (`X_test`) berdasarkan model yang telah dilatih.

4. **`acc_rf = accuracy_score(y_test, y_pred_rf)`**:  
   Menghitung akurasi prediksi model pada *testing set* dengan membandingkan `y_pred_rf` (prediksi) dan `y_test` (label asli).

5. **`print(...)`**:  
   Menampilkan akurasi model *Random Forest* dalam format desimal (2 tempat) dan nilai akurasinya.

# Praktikum 2 Boosting dengan AdaBoost

Import Library

In [8]:
import numpy as np
import pandas as pd
from sklearn.tree import DecisionTreeClassifier # import DT
from sklearn.ensemble import AdaBoostClassifier # import AdaBoost
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import LabelEncoder # Kebutuhan encoding label

   - `numpy` dan `pandas`: Untuk manipulasi data.
   - `DecisionTreeClassifier`: Model *Decision Tree* untuk klasifikasi.
   - `AdaBoostClassifier`: Model *AdaBoost* untuk meningkatkan akurasi.
   - `train_test_split`: Untuk membagi data menjadi *training* dan *testing* set.
   - `accuracy_score`, `classification_report`: Untuk evaluasi performa model.
   - `LabelEncoder`: Untuk mengubah label kategori menjadi numerik.

Persiapan Data

In [10]:
# Load data
df = pd.read_csv('/content/drive/MyDrive/dataset/iris.csv')

df.head()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa

1. **`df = pd.read_csv('/content/drive/MyDrive/dataset/iris.csv')`**:  
   Membaca file CSV (`iris.csv`) dan menyimpannya ke dalam DataFrame `df`.

2. **`df.head()`**:  
   Menampilkan 5 baris pertama dari DataFrame `df` untuk memberikan gambaran awal tentang struktur dan isi data.

In [11]:
# Cek kolom null
df.isnull().sum()

Id               0
SepalLengthCm    0
SepalWidthCm     0
PetalLengthCm    0
PetalWidthCm     0
Species          0
dtype: int64

1. **`df.isnull().sum()`**:  
   - Mengecek setiap kolom dalam DataFrame untuk nilai `NaN` (null/missing values).
   - Menghitung jumlah nilai yang hilang di setiap kolom.

In [12]:
# Seleksi fitur
X = df.iloc[:,2:-1]
y = df['Species']

# encode label
ec = LabelEncoder()
y = ec.fit_transform(y)

# Cek jumlah fitur dan instance
print(X.shape)

# Cek label
print(y)

(150, 3)
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


1. **`X = df.iloc[:, 2:-1]`**:  
   Mengambil subset DataFrame `df` mulai dari kolom ke-3 hingga kolom sebelum yang terakhir sebagai fitur (`X`).

2. **`y = df['Species']`**:  
   Mengambil kolom `Species` sebagai label (`y`).

3. **`ec = LabelEncoder()`**:  
   Inisialisasi objek *LabelEncoder* untuk mengubah label kategori menjadi numerik.

4. **`y = ec.fit_transform(y)`**:  
   Mengubah label `y` (spesies) menjadi bentuk numerik menggunakan *LabelEncoder*.

5. **`print(X.shape)`**:  
   Menampilkan jumlah fitur (kolom) dan instance (baris) dalam `X`.

6. **`print(y)`**:  
   Menampilkan label yang telah di-encode untuk memastikan bahwa pengkodean telah dilakukan dengan benar.

Split data training dan testing

In [13]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

1. **`train_test_split(X, y, test_size=0.2, random_state=1)`**:  
   - Membagi data fitur (`X`) dan label (`y`) menjadi dua bagian:
     - 80% digunakan sebagai *training set*.
     - 20% digunakan sebagai *testing set*.
   - `random_state=1`: Mengatur *random seed* agar hasil pembagian tetap konsisten setiap kali dijalankan.

2. **`X_train, X_test, y_train, y_test`**:  
   - Variabel ini menyimpan data yang dibagi:
     - `X_train` dan `y_train`: Data untuk melatih model.
     - `X_test` dan `y_test`: Data untuk menguji performa model.

Training Decision Tree

In [14]:
# Secara default, DecisionTreeClassifier dari scikit-learn akan menggunakan nilai "Gini" untuk kriteria
# Terdapat beberapa "hyperparamater" yang dapat digunakan. Silahka baca dokumentasi
# Pada kasus ini kita akan menggunakan parameter default
dt = DecisionTreeClassifier()

# Sesuaikan dt ke set training
dt.fit(X_train, y_train)

# Memprediksi label set test
y_pred_dt = dt.predict(X_test)

#  menghitung set accuracy
acc_dt = accuracy_score(y_test, y_pred_dt)
print("Test set accuracy: {:.2f}".format(acc_dt))
print(f"Test set accuracy: {acc_dt}")

Test set accuracy: 0.97
Test set accuracy: 0.9666666666666667


1. **`dt = DecisionTreeClassifier()`**:  
   - Inisialisasi model *Decision Tree* dengan parameter default (menggunakan kriteria Gini).

2. **`dt.fit(X_train, y_train)`**:  
   - Melatih model *Decision Tree* menggunakan *training set* (`X_train`, `y_train`).

3. **`y_pred_dt = dt.predict(X_test)`**:  
   - Memprediksi label untuk *testing set* (`X_test`) berdasarkan model yang telah dilatih.

4. **`acc_dt = accuracy_score(y_test, y_pred_dt)`**:  
   - Menghitung akurasi prediksi model pada *testing set* dengan membandingkan `y_pred_dt` (prediksi) dan `y_test` (label asli).

5. **`print(...)`**:  
   - Menampilkan akurasi model dalam format desimal (2 tempat) dan nilai akurasinya.

Training AdaBoost

In [15]:
# Pada kasus kali ini kita akan menggunakan estimator pada AdaBoost
# Untuk detail parameter (hyperparameter) silahkan cek dokumentasi

ada = AdaBoostClassifier(n_estimators=2)

# Sesuaikan dt ke set training
ada.fit(X_train, y_train)

# Memprediksi label set test
y_pred_ada = ada.predict(X_test)

#  menghitung set accuracy
acc_ada = accuracy_score(y_test, y_pred_ada)
print("Test set accuracy: {:.2f}".format(acc_ada))
print(f"Test set accuracy: {acc_ada}")

Test set accuracy: 0.97
Test set accuracy: 0.9666666666666667


/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_weight_boosting.py:527: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(


1. **`ada = AdaBoostClassifier(n_estimators=2)`**:  
   - Inisialisasi model *AdaBoost* dengan jumlah estimator (pohon keputusan) sebanyak 2.

2. **`ada.fit(X_train, y_train)`**:  
   - Melatih model *AdaBoost* menggunakan *training set* (`X_train`, `y_train`).

3. **`y_pred_ada = ada.predict(X_test)`**:  
   - Memprediksi label untuk *testing set* (`X_test`) berdasarkan model yang telah dilatih.

4. **`acc_ada = accuracy_score(y_test, y_pred_ada)`**:  
   - Menghitung akurasi prediksi model pada *testing set* dengan membandingkan `y_pred_ada` (prediksi) dan `y_test` (label asli).

5. **`print(...)`**:  
   - Menampilkan akurasi model *AdaBoost* dalam format desimal (2 tempat) dan nilai akurasinya.

# Praktikum 3 Stacking

Menggunakan data wbc.csv

In [23]:
import pandas as pd
from sklearn.ensemble import RandomForestClassifier, StackingClassifier, AdaBoostClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.impute import SimpleImputer
from sklearn.metrics import accuracy_score

# Memuat dataset
wbc_data = pd.read_csv('/content/drive/MyDrive/dataset/wbc (1).csv')

# Hapus kolom yang tidak memiliki nilai
wbc_data = wbc_data.drop(columns=['Unnamed: 32'], errors='ignore')  # Hapus kolom Unnamed: 32 jika ada

# Proses untuk dataset WBC
X = wbc_data.drop('diagnosis', axis=1)  # Ganti dengan nama kolom yang sesuai
y = wbc_data['diagnosis']  # Ganti dengan nama kolom yang sesuai

# Mengganti nilai NaN dengan median kolom
imputer = SimpleImputer(strategy='median')
X_imputed = imputer.fit_transform(X)

# Membagi dataset WBC menjadi data latih dan data uji
X_train, X_test, y_train, y_test = train_test_split(X_imputed, y, stratify=y, random_state=42)

# Model Random Forest
rf_model = RandomForestClassifier(n_estimators=50, random_state=42)
rf_model.fit(X_train, y_train)
rf_accuracy = rf_model.score(X_test, y_test)

# Model AdaBoost
ada_model = AdaBoostClassifier(n_estimators=50, random_state=42)
ada_model.fit(X_train, y_train)
ada_accuracy = ada_model.score(X_test, y_test)

# Model Stacking
layer_one_estimators = [
    ('rf_1', RandomForestClassifier(n_estimators=10, random_state=42)),
    ('knn_1', KNeighborsClassifier(n_neighbors=5))
]
layer_two_estimators = [
    ('dt_2', DecisionTreeClassifier()),
    ('rf_2', RandomForestClassifier(n_estimators=50, random_state=42)),
]
layer_two = StackingClassifier(estimators=layer_two_estimators, final_estimator=LogisticRegression())

stacking_model = StackingClassifier(estimators=layer_one_estimators, final_estimator=layer_two)
stacking_model.fit(X_train, y_train)
stacking_accuracy = stacking_model.score(X_test, y_test)

# Menampilkan akurasi
print(f'Akurasi Random Forest: {rf_accuracy:.4f}')
print(f'Akurasi Adaboost: {ada_accuracy:.4f}')
print(f'Akurasi Stacking: {stacking_accuracy:.4f}')

# Menghitung perbedaan akurasi
print(f'Perbedaan akurasi antara Random Forest dan Adaboost: {rf_accuracy - ada_accuracy:.4f}')
print(f'Perbedaan akurasi antara Random Forest dan Stacking: {rf_accuracy - stacking_accuracy:.4f}')
print(f'Perbedaan akurasi antara Adaboost dan Stacking: {ada_accuracy - stacking_accuracy:.4f}')

/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_weight_boosting.py:527: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(


Akurasi Random Forest: 0.9510
Akurasi Adaboost: 0.9650
Akurasi Stacking: 0.9580
Perbedaan akurasi antara Random Forest dan Adaboost: -0.0140
Perbedaan akurasi antara Random Forest dan Stacking: -0.0070
Perbedaan akurasi antara Adaboost dan Stacking: 0.0070


### Kesimpulan
Berdasarkan hasil pengujian model:

- **Akurasi**:
  - Random Forest: **0.9510**
  - Adaboost: **0.9650**
  - Stacking: **0.9580**

- **Perbedaan Akurasi**:
  - Antara Random Forest dan Adaboost: **-0.0140** (Adaboost lebih baik)
  - Antara Random Forest dan Stacking: **-0.0070** (Stacking lebih baik)
  - Antara Adaboost dan Stacking: **0.0070** (Adaboost lebih baik)

Model Adaboost memiliki akurasi tertinggi, diikuti oleh Stacking dan Random Forest. Meskipun ketiga model menunjukkan kinerja yang baik, Adaboost terbukti lebih efektif dalam mengklasifikasikan data pada tugas ini.

Menggunakan data iris.csv

In [21]:
import pandas as pd
from sklearn.datasets import load_iris
from sklearn.ensemble import RandomForestClassifier, StackingClassifier, AdaBoostClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier

# Load dataset Iris
iris = load_iris()
X = iris.data  # fitur
y = iris.target  # label

# Split dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, random_state=42)

# Define estimators untuk Stacking
layer_one_estimators = [
    ('rf', RandomForestClassifier(n_estimators=10, random_state=42)),
    ('knn', KNeighborsClassifier(n_neighbors=5))
]

layer_two_estimators = [
    ('dt', DecisionTreeClassifier()),
    ('rf_2', RandomForestClassifier(n_estimators=50, random_state=42)),
]

# Model Stacking
layer_two = StackingClassifier(estimators=layer_two_estimators, final_estimator=LogisticRegression())
clf_stacking = StackingClassifier(estimators=layer_one_estimators, final_estimator=layer_two)

# Model Random Forest
rf_model = RandomForestClassifier(n_estimators=50, random_state=42)
rf_model.fit(X_train, y_train)
rf_accuracy = rf_model.score(X_test, y_test)

# Model Adaboost
ada_model = AdaBoostClassifier(n_estimators=50, random_state=42)
ada_model.fit(X_train, y_train)
ada_accuracy = ada_model.score(X_test, y_test)

# Model Stacking
clf_stacking.fit(X_train, y_train)
stacking_accuracy = clf_stacking.score(X_test, y_test)

# Tampilkan hasil akurasi
print(f"Akurasi Random Forest: {rf_accuracy:.4f}")
print(f"Akurasi Adaboost: {ada_accuracy:.4f}")
print(f"Akurasi Stacking: {stacking_accuracy:.4f}")

# Hitung perbedaan akurasi
print(f"Perbedaan Akurasi antara Random Forest dan Adaboost: {rf_accuracy - ada_accuracy:.4f}")
print(f"Perbedaan Akurasi antara Random Forest dan Stacking: {rf_accuracy - stacking_accuracy:.4f}")
print(f"Perbedaan Akurasi antara Adaboost dan Stacking: {ada_accuracy - stacking_accuracy:.4f}")


/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_weight_boosting.py:527: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(


Akurasi Random Forest: 0.9211
Akurasi Adaboost: 0.8947
Akurasi Stacking: 0.9737
Perbedaan Akurasi antara Random Forest dan Adaboost: 0.0263
Perbedaan Akurasi antara Random Forest dan Stacking: -0.0526
Perbedaan Akurasi antara Adaboost dan Stacking: -0.0789


### Kesimpulan

Berdasarkan hasil pengujian model:

**Akurasi:**
- Random Forest: 0.9211
- Adaboost: 0.8947
- Stacking: 0.9737

**Perbedaan Akurasi:**
- Antara Random Forest dan Adaboost: 0.0263 (Random Forest lebih baik)
- Antara Random Forest dan Stacking: -0.0526 (Stacking lebih baik)
- Antara Adaboost dan Stacking: -0.0789 (Stacking lebih baik)

Model Stacking memiliki akurasi tertinggi, diikuti oleh Random Forest dan Adaboost. Meskipun Random Forest menunjukkan kinerja yang lebih baik dibandingkan Adaboost, Stacking terbukti lebih efektif dalam mengklasifikasikan data pada tugas ini, dengan perbedaan akurasi yang signifikan antara Stacking dan kedua model lainnya.

# Praktikum 4 Stacking dengan Voting

Import Library

In [24]:
import numpy as np
import pandas as pd
from sklearn.naive_bayes import GaussianNB # import Naive Bayes model Gaussian (asumsi data terdistribusi normal)
from sklearn.svm import SVC # import SVM classifier
from sklearn.ensemble import VotingClassifier # import model Voting
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report

Kode ini mengimpor pustaka yang diperlukan untuk membangun model klasifikasi menggunakan **Naive Bayes** dan **SVM**. **VotingClassifier** akan menggabungkan kedua model untuk meningkatkan akurasi. Dataset akan dibagi menjadi data pelatihan dan pengujian, dan metrik evaluasi seperti **akurasi** dan **laporan klasifikasi** akan digunakan untuk menilai kinerja model.

Persiapan Data

In [25]:
# Load Data

dbt = pd.read_csv('/content/drive/MyDrive/dataset/diabetes.csv')

dbt.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

Kode ini memuat dataset diabetes dari file CSV yang terletak di path yang ditentukan. Fungsi `pd.read_csv()` digunakan untuk membaca file CSV dan menyimpannya ke dalam variabel `dbt`. Kemudian, `dbt.head()` menampilkan lima baris pertama dari dataset, yang memungkinkan pengguna untuk melihat struktur dan beberapa contoh data yang terdapat dalam dataset tersebut.

In [26]:
# Cek nama kolom
dbt.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

Kode ini digunakan untuk memeriksa nama-nama kolom dalam dataset diabetes yang telah dimuat sebelumnya. Dengan menggunakan `dbt.columns`, Anda dapat melihat semua kolom yang tersedia dalam DataFrame `dbt`. Ini membantu untuk memahami fitur yang ada dalam dataset dan memudahkan analisis atau pemodelan lebih lanjut.

In [27]:
# Cek kolom null
dbt.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

Kode ini digunakan untuk memeriksa apakah ada nilai kosong (null) dalam setiap kolom dari DataFrame `dbt`. Dengan menggunakan `dbt.isnull().sum()`, Anda akan mendapatkan jumlah nilai null untuk setiap kolom. Ini penting untuk memahami kualitas data, karena nilai kosong dapat mempengaruhi analisis dan hasil model. Jika ditemukan nilai null, langkah-langkah tambahan seperti imputasi atau penghapusan mungkin diperlukan sebelum melanjutkan ke proses pemodelan.

In [28]:
# Pada kasus ini, agak tidak masuk akal jika beberapa parameter bernilai 0
# sebagai contoh adalah nilai 'Glucose', 'BloodPlessure' ataupun 'Insulin'.
# Sekecil apapun nilainya, setiap manusia yang hidup pasti miliki nilai-nilai tersebut

# Kita akan manipulasi nilai yang 0 dengan melakukan 'imputasi' atau mengganti nilainya dengan nilai sintetis
# Pada kasus ini, kita akan menggunakan nilai mean

# Cek kolom neng nilai 0
feature_columns = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age']
for column in feature_columns:
    print("============================================")
    print(f"{column} ==> Missing zeros : {len(dbt.loc[dbt[column] == 0])}")

Pregnancies ==> Missing zeros : 111
Glucose ==> Missing zeros : 5
BloodPressure ==> Missing zeros : 35
SkinThickness ==> Missing zeros : 227
Insulin ==> Missing zeros : 374
BMI ==> Missing zeros : 11
DiabetesPedigreeFunction ==> Missing zeros : 0
Age ==> Missing zeros : 0


1. **List Kolom**: `feature_columns` berisi daftar kolom yang ingin diperiksa.
2. **Looping**: Menggunakan loop `for` untuk iterasi setiap kolom dalam `feature_columns`.
3. **Print Statement**: Untuk setiap kolom, program mencetak jumlah nilai nol yang ditemukan. Fungsi `len(dbt.loc[dbt[column] == 0])` digunakan untuk menghitung jumlah baris di mana nilai kolom tersebut sama dengan nol.


In [29]:
# Impute nilai 0 dengan mean
from sklearn.impute import SimpleImputer

fill_values = SimpleImputer(missing_values=0, strategy="mean", copy=False)

dbt[feature_columns] = fill_values.fit_transform(dbt[feature_columns])

1. **Import SimpleImputer**: `from sklearn.impute import SimpleImputer` mengimpor kelas `SimpleImputer` dari pustaka Scikit-learn, yang digunakan untuk mengimputasi (mengisi) nilai yang hilang dalam dataset.

2. **Mendefinisikan Imputer**:
   - `missing_values=0`: Menunjukkan bahwa nilai yang akan diimputasi adalah nilai nol.
   - `strategy="mean"`: Menyatakan bahwa nilai nol akan diganti dengan nilai rata-rata dari kolom yang bersangkutan.
   - `copy=False`: Mengizinkan perubahan dilakukan pada DataFrame asli, bukan membuat salinan baru.

3. **Mengimputasi Nilai Nol**:
   - `fit_transform(dbt[feature_columns])`: Metode ini menghitung nilai rata-rata untuk kolom yang ditentukan dan menggantikan nilai nol dengan rata-rata tersebut.
   - Hasilnya kemudian disimpan kembali ke DataFrame `dbt` pada kolom yang sama.

Split data training dan testing

In [30]:
X = dbt[feature_columns]
y = dbt.Outcome

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

1. **Mendefinisikan Fitur dan Target**:
   - `X` berisi fitur dari dataset `dbt` berdasarkan `feature_columns`.
   - `y` adalah kolom `Outcome`, yang merupakan label target.

2. **Membagi Dataset**:
   - `train_test_split` membagi `X` dan `y` menjadi set pelatihan (70%) dan set pengujian (30%).
   - `random_state=42` memastikan pembagian data dapat direproduksi.

Training dengan GaussianNB
- Standarisasi Fitur

In [31]:
# Karena asumsi Gaussian NB adalah data terdistribusi secara normal,
# maka kita perlu melakukan standarisasi

from sklearn.preprocessing import StandardScaler

sc = StandardScaler()

# Standarisasi pada fitur di X_train dan X_test
X_train_std = sc.fit_transform(X_train)
X_test_std = sc.transform(X_test)

1. **Standarisasi Data**:
   - Menggunakan `StandardScaler` dari `sklearn.preprocessing`, kode ini menyiapkan proses standarisasi untuk fitur. Standarisasi mengubah data sehingga memiliki rata-rata 0 dan deviasi standar 1, yang sesuai dengan asumsi model Gaussian Naive Bayes.

2. **Menerapkan Standarisasi**:
   - `sc.fit_transform(X_train)` menghitung parameter (rata-rata dan deviasi standar) dari `X_train` dan menerapkannya untuk standarisasi.
   - `sc.transform(X_test)` menerapkan parameter yang sama pada `X_test` untuk memastikan data pengujian distandarisasi dengan cara yang sama seperti data pelatihan.

- Training dan Evaluasi

In [32]:
# Buat obyek GaussianNB
gnb_std = GaussianNB()

# Fit dengan data yang telah di standarisasi
gnb_std.fit(X_train_std, y_train)

# Prediksi dengan data test
y_pred_gnb = gnb_std.predict(X_test_std)

# Evaluasi akurasi testing data
acc_gnb = accuracy_score(y_test, y_pred_gnb)

# Print hasil evaluasi
print("Test set accuracy: {:.2f}".format(acc_gnb))
print(f"Test set accuracy: {acc_gnb}")

Test set accuracy: 0.74
Test set accuracy: 0.7359307359307359


Gaussian Naive Bayes (GNB):
- Akurasi: 0.74
- Model ini memberikan hasil yang baik dengan asumsi distribusi data mengikuti distribusi normal.

Training dengan SVM Linier

In [33]:
# Model SVM linier tanpa tunnning hyperparameter
svm_lin = SVC(kernel='linear')

# Fit ke model
svm_lin.fit(X_train_std, y_train)

# Prediksi
y_pred_svm_lin = svm_lin.predict(X_test_std)

# Evaluasi akurasi testing data
acc_svm_lin = accuracy_score(y_test, y_pred_svm_lin)

# Print hasil evaluasi
print("Test set accuracy: {:.2f}".format(acc_svm_lin))
print(f"Test set accuracy: {acc_svm_lin}")

Test set accuracy: 0.74
Test set accuracy: 0.7402597402597403


Support Vector Machine (SVM) dengan kernel linear:
- Akurasi: 0.74
- SVM linear menunjukkan kinerja yang sebanding dengan GNB, menunjukkan bahwa model ini juga efektif dalam klasifikasi.

Training dengan SVM RBF

In [34]:
# Model SVM RBF tanpa tunnning hyperparameter
svm_rbf = SVC(kernel='rbf')

# Fit ke model
svm_rbf.fit(X_train_std, y_train)

# Prediksi
y_pred_svm_rbf = svm_rbf.predict(X_test_std)

# Evaluasi akurasi testing data
acc_svm_rbf = accuracy_score(y_test, y_pred_svm_rbf)

# Print hasil evaluasi
print("Test set accuracy: {:.2f}".format(acc_svm_rbf))
print(f"Test set accuracy: {acc_svm_rbf}")

Test set accuracy: 0.72
Test set accuracy: 0.7229437229437229


Support Vector Machine (SVM) dengan kernel RBF:
- Akurasi: 0.72
- Meskipun hasilnya sedikit lebih rendah dibandingkan dengan GNB dan SVM linear, model ini tetap memberikan performa yang baik.

Training dengan Voting

In [35]:
# Definisikan algoritma yang akan digunakan untuk voting

clf1 = GaussianNB()
clf2 = SVC(kernel='linear')
clf3 = SVC(kernel='rbf', probability=True)

# model hard voting
voting = VotingClassifier(estimators=[('GaussianNB', clf1), ('SVM-LIN', clf2), ('SVM-RBF', clf3)], voting='hard')

# Fit model
voting.fit(X_train_std, y_train)

# Prediksi
y_pred_vt1 = voting.predict(X_test_std)

# Evaluasi akurasi testing data
acc_vt1 = accuracy_score(y_test, y_pred_vt1)

# Print hasil evaluasi
print('Voting Hard')
print("Test set accuracy: {:.2f}".format(acc_vt1))
print(f"Test set accuracy: {acc_vt1}")

Voting Hard
Test set accuracy: 0.74
Test set accuracy: 0.7402597402597403


Voting Classifier (Hard Voting):
- Akurasi: 0.74
- Model Voting menggabungkan prediksi dari ketiga algoritma sebelumnya, yaitu GNB dan kedua jenis SVM. Hasil akurasi yang dihasilkan serupa dengan model GNB dan SVM linear, menunjukkan bahwa kombinasi ini memberikan stabilitas yang lebih baik dan mengurangi risiko kesalahan yang diakibatkan oleh satu model saja.

Secara keseluruhan, semua model menunjukkan kinerja yang memuaskan dengan akurasi yang konsisten di sekitar 0.74. Pendekatan Voting Classifier dapat dianggap sebagai metode yang efektif dalam meningkatkan akurasi klasifikasi, dengan memanfaatkan kelebihan dari masing-masing algoritma yang digunakan. Dengan demikian, baik model individual maupun model ensemble dapat diterapkan dengan baik untuk masalah klasifikasi pada dataset diabetes ini.